### ISSUE TITLE
KillUnusedSegments ignores segments that are marked unused after it last ran

### ISSUE THREAD

**Issue by JRobTS** [apache/druid#15912]
When a retention policy of a datasource is greater than the last coordinator issued kill for that datasource, it will not issue kill commands for older segments. This is a regression from previous versions of Druid. 

This has the effect that deep storage can grow unbounded which is a problem for people who have limited storage or pay for storage.

### Affected Version

28.0.1

### Description

Suppose the Druid kill configuration is:
```
druid.coordinator.kill.on=true
druid.coordinator.kill.period=PT1H
druid.coordinator.kill.durationToRetain=P1D
druid.coordinator.kill.bufferPeriod=PT6H
druid.coordinator.kill.maxSegments=10000
```

And we have one or more datasources with a retention policy like:
```
loadByPeriod P7D
dropForever
```

The Druid Coordinator will issue kill tasks but only for segments that are older than the last run of KillUnusedSegments. It will not issue kill tasks for segments that are older than the last run which means that the Deep Storage will grow endlessly instead of Druid cleaning up segments that are older than the last run of the KillUnusedSegments.

The task should start with the oldest segments per datasource, and work its way forward. Any segment marked as unused (used=0) in druid_segments metadata table should be fair game.

For further details including debugging, see Slack discussion here:
https://apachedruidworkspace.slack.com/archives/C0309C9L90D/p1707931446776669

**Comment by JRobTS** [comment]
So it appears that this sometimes works and I traced the reason why it sometimes works:

After a leader election it appears the coordinator is working from a clean slate and is able to delete the older segments. So I guess a side-effect of leader election is that it will delete old segments but while a coordinator holds leadership it will not delete old segments.

**Comment by zachjsh** [comment]
@JRobTS , thanks for submitting this issue. As I recall, the coordinator should loop around to the older segments that i missed on way up, after it  has found no new unused segments for a given later interval. Are you finding this not to be the case?

**Comment by JRobTS** [comment]
That's correct. Old unused segments are not included in the interval unless
a leader election just happened. To be fair, the kill job does find unused
segments, it's just finding the recent stuff from compaction and not
including the older unused segments from drop forever rule.

**Comment by JRobTS** [comment]
We are using the default indexingPeriod of 30 minutes and we have `druid.coordinator.kill.period=PT1H` (although the kill actually runs every 1.5 hours, I've logged a separate ticket on that).

Compaction is definitely occurring more frequently than the kill operation but (from a requirements perspective) this shouldn't stop kill from deleting the unused segments created by the dropForever rule. I have set `druid.coordinator.kill.maxSegments=10000` but it only gets around 100 or so segments per run.

I will enable continuous auto-kill by removing `druid.coordinator.kill.period` and monitor.

**Comment by JRobTS** [comment]
I have not testing on the latest version of Druid but I believe this issue is still relevant for any indexingPeriod >= compaction period.
